compiler/rustc_index/src/interval.rs

```diff
@@ -140,6 +140,52 @@ impl<I: Idx> IntervalSet<I> {
         result
     }
 
+    /// Specialized version of `insert` when we know that the inserted point is *before* any
+    /// contained.
+    pub fn prepend(&mut self, point: I) {
+        let point = point.index() as u32;
+
+        if let Some((first_start, _)) = self.map.first_mut() {
+            assert!(point <= *first_start);
+            if point == *first_start {
+                // The point is already present in the set.
+            } else if point + 1 == *first_start {
+                // Just extend the first range.
+                *first_start = point;
+            } else {
+                self.map.insert(0, (point, point));
```

Comment:
It's unfortunate that this is O(n^2). It is possible to make this O(n) by building the intervals in reverse order and then reversing the whole array at the end. Actually even the final reversal isn't strictly needed if you take the reversed order into account when unioning rows and checking for disjointness.